### PR TITLE
Fix tf.matching_files issue with access denied subdirectory.

### LINKED ISSUE #19274: tf.matching_files fails when there is a file with wrong permissions in a subdirectory
### System information
- **Have I written custom code (as opposed to using a stock example script provided in TensorFlow)**: yes
- **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**: Ubuntu18.04
- **TensorFlow installed from (source or binary)**: pip installed binary
- **TensorFlow version (use command below)**: 1.8
- **Python version**: 3.6
- **Bazel version (if compiling from source)**: NA
- **GCC/Compiler version (if compiling from source)**: NA
- **CUDA/cuDNN version**: NA
- **GPU model and memory**: NA
- **Exact command to reproduce**: see below

```python
import tensorflow as tf
from tensorflow.python.framework.graph_util import convert_variables_to_constants

graph = tf.Graph()
with graph.as_default():
    output1 = tf.matching_files('/some/dir', name='output1', )
    sess = tf.InteractiveSession(graph=graph)
    print(sess.run([output1]))
```

### Describe the problem

This simple script fails if some dir contains a file without read access somewhere in subfolders of `/some/dir`. So for instance if I don't have permission to read `/some/dir/subfolder/filebelongingtosomeoneelse`, the script fails (while I have permissions for subfolder). 
Strange thing is that neither read access needed to know its filename, nor it is actually will be listed in the output (so it should be skipped during scanning).

### PR DESCRIPTION
**Pull request by yongtang**
This fix tries to address the issue raised in #19274 where
tf.matching_files will return an error if no permission for
subdirectories (Even if the subdirectories does not match).

This fix addresses the issue by ignore the case for `permisson denied`
during the directory traversal.

This fix fixes #19274.

Signed-off-by: Yong Tang <yong.tang.github@outlook.com>

### CODE REVIEW

**Comment by tensorflowbutler** [comment]
Nagging Assignee @rohan100jain: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @rohan100jain: It has been 29 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @rohan100jain: It has been 59 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @rohan100jain: It has been 74 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

File: tensorflow/core/platform/file_system_helper.cc [hunk 1]
```diff
@@ -82,7 +82,10 @@ Status GetMatchingPaths(FileSystem* fs, Env* env, const string& pattern,
     dir_q.pop_front();
     std::vector<string> children;
     Status s = fs->GetChildren(current_dir, &children);
-    ret.Update(s);
+    // We will ignore permission denied error, and update status otherwise.
+    if (s.code() != tensorflow::error::PERMISSION_DENIED) {
```
Comment (rohan100jain) [inline tensorflow/core/platform/file_system_helper.cc]: Hmmm so the semantics of GetChildren are not clear when PERMISSION_DENIED is returned. Should we try to process the children in the vector? Might be better to bail if we hit PERMISSION_DENIED?

**Review by rohan100jain — CHANGES_REQUESTED** [review]
Also, can you add a test to file_io_test.py. Create a directory structure and use chmod to change file permissions so that you don't have access to a sub directory. Make sure that you can traverse the other files.

**Comment by yifeif** [comment]
@yongtang are you still working on this PR?

**Comment by yongtang** [comment]
@rohan100jain @yifeif Thanks for the help. The PR has been updated with test cases added. Also, when GetChildren encountered PERMISSION_DENIED will pass. Please take a look and let me know if there are any issues.

File: tensorflow/python/lib/io/file_io_test.py [hunk 2]
```diff
@@ -582,5 +584,31 @@ def testFileCrc32Binary(self):
     self.assertTrue(crc1 != crc2)
     self.assertEqual(crc2, crc3)
 
+  def testMatchingFilesPermission(self):
+    # Test case for GitHub issue 19274.
```
Comment (rohan100jain) [inline tensorflow/python/lib/io/file_io_test.py]: Remove this comment line

File: tensorflow/python/lib/io/file_io_test.py [hunk 3]
```diff
@@ -582,5 +584,31 @@ def testFileCrc32Binary(self):
...
+      file_path = os.path.join(any_path, name)
+      file_io.FileIO(file_path, mode="w").write("testing")
+    file_path = os.path.join(noread_path, "file4.txt")
+    file_io.FileIO(file_path, mode="w").write("testing")
+    # Change noread to noread access.
+    os.chmod(noread_path, 0)
+    expected_match = [compat.as_bytes(dir_path)]
+    with self.test_session() as sess:
+      self.assertItemsEqual(
+          gen_io_ops.matching_files(dir_path).eval(), expected_match)
```
Comment (rohan100jain) [inline tensorflow/python/lib/io/file_io_test.py]: why use gen_io_ops.matching_files.. why not file_io.get_matching_files instead? In fact try looking for .../test_dir/*/file*.txt and expect to only find file1.txt, file2.txt and file3.txt

**Comment by rohan100jain** [comment]
Thanks @yongtang!

**Comment by yongtang** [comment]
Thanks @rohan100jain  for the review. The PR has been updated. Please take a look.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @rohan100jain: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @rohan100jain: It has been 29 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @rohan100jain: It has been 44 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Review by rohan100jain — APPROVED** [review]

**Comment by yongtang** [comment]
@yifeif The `feedback/copybara` is failing but I could not see the details. Is there anyway to see what might be the error in `feedback/copybara`?

**Pull request merged**